### PR TITLE
build: validate `build-system` table schema

### LINKED ISSUE #364: Validate build-system table schema
I just spend a bunch of time debugging an error due to me mistakenly setting `build-backend = ['something']` in the `build-system` table.

We should verify the schema (types) and provide a helpful error message if something is wrong.

https://www.python.org/dev/peps/pep-0518/#build-system-table
https://www.python.org/dev/peps/pep-0518/#json-schema

### PR DESCRIPTION
**Pull request by layday**
Closes #364.

### CODE REVIEW

**Review by gaborbernat — APPROVED** [review]

File: src/build/__init__.py
Comment (FFY00) [inline src/build/__init__.py:103]:
```suggestion
        return f'Failed to validate `build-system` in pyproject.toml: {self.args[0]}'
```

File: tests/test_projectbuilder.py
Comment (FFY00) [inline tests/test_projectbuilder.py:576]: Can we parametrize this test instead?

**Review by henryiii — APPROVED** [review]
Once existing comments are addressed, LGTM.

File: tests/test_projectbuilder.py
Comment (layday) [reply tests/test_projectbuilder.py:576]: I would've done but it's kind of awkward when mixing test cases which raise.

Comment (layday) [reply tests/test_projectbuilder.py]: done

Comment (layday) [reply tests/test_projectbuilder.py]: The new parametrised tests fail unless I exclude `test_toml_instead_of_tomli`...

File: tests/test_projectbuilder.py
Comment (FFY00) [inline tests/test_projectbuilder.py:587]: We can just parameterize the match.

https://github.com/FFY00/python-pep621/blob/f374fa2fe8f53fc3b24ef9e08b00870fef8fce30/tests/test_standard_metadata.py#L479

File: tests/test_projectbuilder.py
Comment (FFY00) [inline tests/test_projectbuilder.py:629]: I would rather instantiate a `ProjectBuilder` here.

File: tests/test_projectbuilder.py
Comment (layday) [reply tests/test_projectbuilder.py:587]: Not all of them raise.

Comment (FFY00) [reply tests/test_projectbuilder.py:587]: Then split the test into one for the errors and one for the case it should pass.

Comment (layday) [reply tests/test_projectbuilder.py:587]: Do you not like the way it's structured now or why do you want me to do that?

Comment (FFY00) [reply tests/test_projectbuilder.py:587]: It's to remove duplication and complexity. If you split the test into two you can easily parameterize only the error string.

File: tests/test_projectbuilder.py
Comment (layday) [reply tests/test_projectbuilder.py:629]: I'd have to serialise them all as TOML, create a tmp dir for each and write them on disk - I don't think I can disentangle the disk I/O with the mocker.

File: tests/test_projectbuilder.py
Comment (layday) [reply tests/test_projectbuilder.py]: I hope https://github.com/pypa/build/pull/365/commits/1dd2d353382978d394919cb278b9941510142d66 is acceptable.

File: tests/test_projectbuilder.py
Comment (FFY00) [reply tests/test_projectbuilder.py:626]: I think you can, as you should be able to patch `builtins.open`, but I'll leave that up to you. This is just a nitpick.

Comment (layday) [reply tests/test_projectbuilder.py:626]: I think I might leave that for another day then 😛

File: tests/test_projectbuilder.py
Comment (FFY00) [reply tests/test_projectbuilder.py]: I still think it would be better to split the test into two, and validate the return of `_parse_build_system_table`, but I am not gonna block the PR because of that.

**Review by FFY00 — APPROVED** [review]

File: tests/test_projectbuilder.py
Comment (layday) [reply tests/test_projectbuilder.py]: Done in https://github.com/pypa/build/pull/365/commits/8bab23845d9f002977324a8b928dc1f97fd71c70.

**Pull request merged**
